src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { cloneDeep } from 'lodash';
+
+export interface Permissions {
+  write?: Principals;
+  read?: Principals;
+  management?: Principals;
+  library_read?: Principals;
+  library_write?: Principals;
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export const PERMISSION_TYPE_READ = 'read';
+export const PERMISSION_TYPE_WRITE = 'write';
+export const PERMISSION_TYPE_MANAGEMENT = 'management';
+export const PERMISSION_TYPE_LIBRARY_READ = 'library_read';
+export const PERMISSION_TYPE_LIBRARY_WRITE = 'library_write';
+
+export const PERMISSION_TYPES: Set<string> = new Set([
+  PERMISSION_TYPE_READ,
+  PERMISSION_TYPE_WRITE,
+  PERMISSION_TYPE_MANAGEMENT,
+  PERMISSION_TYPE_LIBRARY_READ,
+  PERMISSION_TYPE_LIBRARY_WRITE,
+]);
+
+const transferPermissionsToMap = (permissions: Permissions) => {
+  const map: Map<string, Principals> = new Map();
+  if (!!permissions.read) {
+    map.set(PERMISSION_TYPE_READ, permissions.read);
+  }
+  if (!!permissions.write) {
+    map.set(PERMISSION_TYPE_WRITE, permissions.write);
+  }
+  if (!!permissions.management) {
+    map.set(PERMISSION_TYPE_MANAGEMENT, permissions.management);
+  }
+  if (!!permissions.library_read) {
+    map.set(PERMISSION_TYPE_LIBRARY_READ, permissions.library_read);
+  }
+  if (!!permissions.library_write) {
+    map.set(PERMISSION_TYPE_LIBRARY_WRITE, permissions.library_write);
+  }
+  return map;
+};
+
+const addPermissionToPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users && !(principals.users?.length === 1 && principals.users[0] === '*')) {
```

Comment:
Is there any possibility to have `principals.users = ['user1', '*', 'user2']`?

Reply:
I think we should avoid this case, when `principals.users` will be updated to `*`, `user1` and `user2` should be deleted.